Crossword

sharabhuiyan opened the pull request:
Created a crossword game:

-  Generate a crossword given input text.
- Created a crossword generation algorithm that finds the first valid crossword
- Each crossword that is generated is different because we randomize the order of the words inputted 
- When a word is found, the text in a row/column turn green
- Crossword contains 3 modes (easy, medium, hard)
    - Easy mode:  definitions and clues are linked and crossed out when word is found
    -  Medium mode: definitions and clues are not linked together, only clues are crossed off when word is found
    - Hard mode: there are extra definitions to throw students off 
    
![image](https://user-images.githubusercontent.com/34693697/111930860-41df2980-8a90-11eb-89ee-9a0b13c43cd3.png)
![image](https://user-images.githubusercontent.com/34693697/111930881-4e638200-8a90-11eb-978b-af78e53af400.png)
![image](https://user-images.githubusercontent.com/34693697/111930889-53283600-8a90-11eb-9467-8110669a059b.png)

JasonLin43212 committed:
Started frontend

JasonLin43212 committed:
Fixed merge conflict

JasonLin43212 committed:
Added backend starter functions:

JasonLin43212 committed:
Added test data and clues

JasonLin43212 committed:
Added letter adding

taokinbo committed:
started making the backbone of the crossword algorithm

taokinbo committed:
A semi-working crossword

JasonLin43212 committed:
Added some styling to crossword and added the rest of the clues

JasonLin43212 committed:
Added correct words

JasonLin43212 committed:
Fixed merge conflict

JasonLin43212 committed:
Added color to the correct cells

JasonLin43212 committed:
Added key navigation

taokinbo committed:
changed crossword algorithm

JasonLin43212 committed:
Added definitions and fill in the blank for clues

JasonLin43212 committed:
Added quiz and crossword links in index page. Fixed quiz

taokinbo committed:
fixed algorithm and returned what is needed for frontend

JasonLin43212 committed:
Started connecting

JasonLin43212 committed:
Improved looks and fixed some bugs with clue starting point

taokinbo committed:
fixed len of word bugs

JasonLin43212 committed:
Fixed bug in frontend with test data

JasonLin43212 committed:
Added give up

JasonLin43212 committed:
Removed comments

JasonLin43212 committed:
Added difficulty and instructions

sharabhuiyan committed:
created docstrings and commented crossword.py

sharabhuiyan committed:
updated docstring

sharabhuiyan committed:
fixed linter errors

did not see need in using enumerate as suggested by linter

taokinbo committed:
fixed linter issue

MBJean commented on backend/app/analysis/crosswords.py:
```diff
@@ -0,0 +1,306 @@
+        indx = random.randint(0, len(all_words) - 1)
+        while all_words[indx] not in words:
+            words.add(all_words[indx])
+    return list(words)
```
This function looks like it does the trick, but two things stand out to me:
1. I think that supplying a `max_num_words` that is greater than or equal to `len(all_words)` will cause an endless loop. Could we guard against that possibility?
2. Regardless of the above, the `random.randint` solution here could create duplicate attempts to insert the same word into `words` and duplicate the lookup of `not in`. This seems unlikely to cause a performance problem given the maximum likely length of `all_words`, but since we're already relying on `random` here, I wonder if we could shortcut some of this manual work by using `random.shuffle` as we do in functions below. i.e., something like:

```
def rand_words(all_words, max_num_words):
  output = all_words.copy()
  random.shuffle(output)
  return output[0:max_num_words]
```

sharabhuiyan replied:
Thank you for the feedback! 
1. In the `get_crossword` function, we do check to see if the` len(all_words)< max_num_words` in this block of code. 
![image](https://user-images.githubusercontent.com/34693697/113050599-0371fa80-9173-11eb-8aa2-4b82f136fb09.png)
2.  You brought up a good point! I have made the changes necessary.

MBJean commented on backend/app/views.py:
```diff
@@ -130,7 +134,7 @@ def update_text(request):
     the id of the text.
     """
-    body = json.loads(request.body.decode('utf-8'))
+    body = json.loads(json.dumps(request.data))
```
If you have a second, could you write out a few words explaining why this change is necessary, and why we need to both `dumps` and `loads` here and on line 123?

JasonLin43212 replied:
The main motivation for the change was that at the time, we were getting a strange error about not being able to read from the `request.body` twice and when I searched up, it was related to Django middleware, which was really confusing.

However, I tested the code using `json.loads(request.body.decode('utf-8'))` now and it seems fine. Should I change it back?

MBJean commented on backend/app/analysis/crosswords.py:
```diff
@@ -0,0 +1,306 @@
+            row = []
+            for j in range(grid_size):
+                row.append(0)
+            grid.append(row)
```
In this function, `grid_size` and `grid` are only used to pass into `make_crossword`, i.e., the assignment of these variables seems relevant to the making of the crossword (what's called in line 282) not the getting of it (the current function). Do you think it would better encapsulate functionality to move the `grid_size` and `grid` assignments/creation into `make_crossword`? Doing so would allow us to get rid of `grid` as an argument for `make_crossword`. This makes some additional sense to me: the `grid` at creation time is simply a product of the `grid_size`, so providing both `grid` and `grid_size` as arguments seems somewhat duplicative.

sharabhuiyan replied:
I agree with this statement. I have made the changes, thanks!

MBJean commented on backend/app/analysis/crosswords.py:
```diff
@@ -0,0 +1,306 @@
+            word_list (list): contains words to be placed into crossword
+            grid_size (int): represents dimension of grid (grid is a square)
+    """
+    first_word = word_list[0]
```
What is the consequence of not sorting the `word_list` by reverse `len(word)`? We perform that sort outside `make_crossword`, so there's the possibility that future us could attempt to use this function again without remembering that it depends on that sort. Would it be worth adding a note in the docstring, or perhaps even performing the sort in this function?

sharabhuiyan replied:
We sorted the `word_list` by `len(word)` because it is easier to place the hardest (longest) word into the crossword because it allows us to have more possibilities for connecting words in the future. I have changed the code so that we perform the sort within `make_crossword`.

MBJean commented on backend/app/analysis/crosswords.py:
```diff
@@ -0,0 +1,306 @@
+    # checks if word is in within grid boundaries
+    return ((direction == "across" and column + word_len <= grid_dim[0] - 1)
+            or (direction == "down" and row + word_len <= grid_dim[1] - 1)) and (
+               row >= 0 and column >= 0)
```
Since `make_crossword` assumes/requires the grid to be a square, could we simplify `grid_dim` as simply `grid_size`?

sharabhuiyan replied:
We decided to use `grid_dim`  in case we were to edit the code in the future and allow other grid shapes.

MBJean replied:
👍

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+            grid[row][col] = curLetter;
+            // Check if a word has been found
+            this.setState({ grid }, this.updateFound);
+        }
```
I'd recommend a refactor of this method. If I read this correctly, we've copied a reference to `this.state.grid` and assigned it to `grid`, but haven't deep copied the value of `this.state.grid` (the 2D array) itself. The consequence is that `grid[row][col] = curLetter` is actually modifying `state.grid[row][col]` without using React's `setState` method. We're then using `setState` to set `state.grid` to itself (since `grid` is just a reference to `state.grid`), which forces the rerender we want to update the view. This presumably works as expected, although we've technically modified the data structure (the 2D array `grid`) twice, only one modification of which will trigger the React rerender. It's been my experience that directly modifying state without using the React `setState` [can cause really ambiguous bugs](https://daveceddia.com/why-not-modify-react-state-directly/) (although admittedly in this case it seems relatively harmless). I think a more React-specific solution would fully swap out the grid for a copy that has had the appropriate 'cell' modified. Something like:

```
const new_grid = [];
const curLetter = e.target.value.toUpperCase();

for (let new_grid_row = 0; new_grid_row < this.state.grid.length; new_grid_row++) {
  const holder = [];
  for (let new_grid_column = 0; new_grid_column < this.state.grid[new_grid_row].length; new_grid_column++) {
    holder.push(this.state.grid[new_grid_row][new_grid_column]);
  }
  new_grid.push(holder);
}

if (ALPHA.includes(curLetter)) {
  new_grid[row][col] = curLetter;
  this.setState({ grid: new_grid }, this.updateFound)
}
```

Or a more functional approach:

```
this.setState(
  (state) => ({
    grid: state.grid.map(
      (temp_row, temp_row_index) => temp_row.map(
        (temp_col, temp_col_index) => row === temp_row_index && col === temp_col_index ? curLetter: temp_col
      )
    )
  }),
  this.updateFound
);
```

One more note: the second code snippet above makes use of the `updater` function pattern described in the [React docs](https://reactjs.org/docs/react-component.html#setstate). The justification for the pattern is outlined well [here](https://medium.com/@wisecobbler/using-a-function-in-setstate-instead-of-an-object-1f5cfd6e55d1). Regardless, the above two code snippets should achieve the same effect.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+            const emptyGrid = data.solution.map((row) => row
+                .map((cell) => (ALPHA.includes(cell) ? '' : '#')));
+            this.setState({
+                crosswordData: data,
```
Did you all consider flattening `crosswordData` into its properties before setting it into the component state? That is: `state.crosswordData` has an unspecified set of properties (I can see from the code below that those properties include `solution`, `clues`, and `definitions`). It looks like the only way to know what those properties are is to consult the API directly to determine the shape of the data. I wonder if it'd be preferable to set expectations around the shape of the return data from that endpoint here, as well. 

We could destructure `response.json()` as:

```
const { solution, clues, definitions } = response.json();
```

So, instead of:

```
state: {
  ...,
  crosswordData: {
    solution: ...,
    clues: ...,
    definitions: ...
   }
}
```

we simply set:

```
state: {
  ...,
  solution: ...,
  clues: ...,
  definitions: ...
}
```

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+    }
+
+    modalHandler = () => {
+        this.setState({ showModal: !this.state.showModal });
```
Please see my comment above about the `updater` function pattern in React's `setState`, and some of the challenges that might occur when referencing state directly as the value of a `setState` update statement.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+    render() {
+        if (!this.state.crosswordData) {
+            return (<LoadingPage />);
+        }
```
While this works, a safer and more explicit pattern pattern would be to introduce a `loading` property to the component's state and key this UI element off of it.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+                        );
+                    }
+                    return '';
+                })
```
I can see this pattern used in a couple of places (i.e., map over an array, guard against the existence of a property, return `''` or `null` as the fallback). We could possibly use the [Array.filter method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) to ensure the array only contains meaningful data and thereby reduce the amount of boilerplate around this pattern.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+                                    onClick={this.modalHandler}>
+                                    <b>?</b>
+                                </button>
+                            </h1>
```
I can see an opportunity to improve the [semantics](https://developer.mozilla.org/en-US/docs/Glossary/Semantics) of this portion of the UI. 

Generally, for the sake of accessibility (particularly screen readers) and of [SEO](https://developer.mozilla.org/en-US/docs/Glossary/SEO), a single rendered HTML page should have a fixed `<h1>` element that [does not change](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements). 

The pattern I see in this code is to swap out the `<h1>` and its content depending on the user's engagement with the difficulty selection buttons. A more conventional pattern might be to render a single, fixed `<h1>` element (with the content being, say, 'Crossword'), and have a `<form>` UI component that contains both a `<label>` describing the user's choices and `<input>` or `<button>` selection options. 

There'd be a whole host of other considerations that would follow from that change (one being: how to announce to a screen reader that there's new content on the page when a user interactions with the difficulty buttons), but we can tackle accessibility more broadly in future PRs. In this case, I think it'd be worth reworking how this `<h1>` element is being modified to take us one step closer toward more semantic HTML.

For those interested in web software development, writing semantic HTML is a foundational skill, so you may find value in skimming through the MDN docs I linked above.

MBJean commented on backend/app/views.py:
```diff
@@ -155,6 +159,43 @@ def delete_text(request):
+            crossword_data['definitions'].append(pos_def[0])
+
+    random.shuffle(crossword_data['definitions'])
+    return Response(crossword_data)
```
Would it be possible to refactor this and push lines 179-195 down into `get_crosswords`? That way we could keep the API layer's responsibilities limited (as we do below with `get_quiz_data`) and more easily unit test the various pieces.

taokinbo replied:
Definition and example setting has now been moved into the get_crosswords functions.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+                found[k].down = correct;
+            }
+        });
+        this.setState({ found });
```
Please see my comment above for the `updateCell` method. In this method, I can see that we're again modifying state twice: once without React's `setState` and once with it.

MBJean commented on frontend/src/crosswordView/crossword.js:
```diff
@@ -0,0 +1,114 @@
+    isCorrect: PropTypes.func,
+};
+
+export default Crossword;
```
I think this component is very well constructed, great work! I appreciate the consistency in the use of variables names throughout this component, though I'd encourage everyone not to be afraid to expand on variable names for the sake of readability. It looks like `r` and `c` in this component are consistently used to refer to something like `row_index` and `column_index`, respectively. It'd probably help our future selves out to expand those variable names to make it clear in everyone context what they refer to. One example: there's a little inconsistency there when we pass up those index references to `this.props.updateCell` in the parent component, which refers to them as `row` and `col`, respectively. I think it'd be worth it to take a quick pass at both this and `crosswordView` to ensure that we're consistent in how we refer to these values.

MBJean commented on frontend/src/crosswordView/crosswordView.scss:
```diff
@@ -0,0 +1,121 @@
+    width: 200px;
+    font-size: 40px;
+
+}
```
The most nitpick of nitpicks: it looks like there's an errant empty line here.

MBJean commented on frontend/src/crosswordView/crosswordView.scss:
```diff
@@ -0,0 +1,121 @@
+
+.definition-title {
+
+}
```
If we don't intend to make use of this className, let's get rid of it.

MBJean commented on frontend/src/crosswordView/crosswordView.scss:
```diff
@@ -0,0 +1,121 @@
+}
+
+$cell-width: 50px;
+$cell-height: 50px;
```
This is a matter of personal style, probably, but I think it's a good idea to pull up SCSS variable assignments to the top of the file so that it's easy for our future selves to find them.

MBJean commented on frontend/src/crosswordView/crosswordView.scss:
```diff
@@ -0,0 +1,121 @@
+    &:focus {
+        outline: none;
+    }
+}
```
This CSS looks well formed! There's a few topics that might be worth discussing, especially if any one has a particular interest in presentation. 

[Using relative units for sizing](https://www.24a11y.com/2019/pixels-vs-relative-units-in-css-why-its-still-a-big-deal/). This is a big aid in responsive design and worth thinking about, if you haven't already.

[Alphabetical ordering for CSS properties](https://medium.com/@jerrylowm/alphabetize-your-css-properties-for-crying-out-loud-780eb1852153). This is a big aid in avoiding headaches when you inevitably need to update old CSS.

Since I'm coming into this well into the project, please feel free to ignore if you all have common patterns that you're following already.

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+                opacity: showModal ? 1 : 0,
+            }}>
+                <div className="modal-header">
+                    <h5 className="modal-title">Instructions</h5>
```
Please see my other comment about semantic HTML. The use of the `<h5>` here violates the [guidelines around HTML headers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#usage_notes) since it doesn't follow in the appropriate header sequence (i.e., sequentially, `<h1>`, `<h2>`, etc.).

MBJean commented on frontend/src/crosswordView/crosswordView.js:
```diff
@@ -0,0 +1,343 @@
+                </div>
+                <Footer/>
+            </>
+        );
```
There's a lot going on in this `render` method, which is making it a bit challenging to grok fully and to skim for things like semantics and accessibility concerns. [This article](https://medium.com/front-end-weekly/react-quick-tip-split-up-your-render-method-5eb30e432e2) outlines a few techniques for keeping your render functions slim, with the general idea being to follow the basic "single-responsibility principle" of programming.

MBJean reviewed (changes requested):
Hi, everyone! I'm picking up the review for this PR. 

Overall, this is fantastic work. There's real a sophistication to how you all have kept each part of the stack responsible for its own concerns. 

I've left a lot of comments below on various places that I can see for improvement or refactor. The goal of most of the suggested refactor is broadly the same: to help our future selves out when we inevitably need to return to this code and continue working on it.

With all of my reviews, please always feel comfortable pushing back or explaining your reasoning. If anyone has any questions or needs clarification, don't hesitate to reach out and we can meet in office hours, the Friday hack session, or schedule some specific time to meet.

One thing that I will consider a blocker: I'd love to see some testing around this work.

And one thing I'm interested in your opinions on: how challenging do you all think it'd be to avoid clues lining up in such a way that they create non-clued, nonsense words (like IPEO and LKSC in the screenshot below)?

![Screen Shot 2021-03-24 at 11 43 33 AM](https://user-images.githubusercontent.com/20114078/112339737-52f98780-8c96-11eb-9283-e7e8f66c237a.png)

JasonLin43212 committed:
Addressed front end issues

sharabhuiyan committed:
addressed PR comments and started resolving clue line-up issue

sharabhuiyan committed:
Merge branch 'crossword' of https://github.com/dhmit/lang_learn into crossword

JasonLin43212 committed:
Fixed python linter errors

taokinbo committed:
gave more restrictions for valid crossword and sorted clues

taokinbo committed:
moved example and definition setting from view to crosswords

taokinbo committed:
fixed linter issues

taokinbo committed:
fixed linters again

taokinbo commented:
@MBJean We have made all relevant changes that were requested and gave reasons for specific changes that were not made under your review comments. We have also looked into the nonsense words that were previously being made and changed our validation code to no longer allow them: 
![image](https://user-images.githubusercontent.com/70776613/113434014-41a72e00-93ae-11eb-850c-ae9733c52611.png)

We eagerly await your rereview :)

taokinbo committed:
removed invalid crosswords

JasonLin43212 committed:
Added tests for crossword

sharabhuiyan commented:
> @MBJean We have made all relevant changes that were requested and gave reasons for specific changes that were not made under your review comments. We have also looked into the nonsense words that were previously being made and changed our validation code to no longer allow them:
> ![image](https://user-images.githubusercontent.com/70776613/113434014-41a72e00-93ae-11eb-850c-ae9733c52611.png)
> 
> We eagerly await your rereview :)

We added tests for the functions that are essential to creating crosswords! Thank you :)

JasonLin43212 committed:
Fixed pylint error

MBJean reviewed (approved):
This looks great! Great job figuring out how to meaningfully test the crossword creation. Merging.
